src/main/java/com/opentable/kafka/util/OffsetMetrics.java

```diff
@@ -0,0 +1,217 @@
+            final Map<Integer, Long> sizes = monitor.getTopicSizes(topic);
+            if (sizes.isEmpty()) {
+                badTopics.add(topic);
+                return;
+            }
+            for (final int part : sizes.keySet()) {
+                builder.put(partitionName(topic, part, "size"), new AtomicLongGauge());
+                builder.put(partitionName(topic, part, "offset"), new AtomicLongGauge());
+                builder.put(partitionName(topic, part, "lag"), new AtomicLongGauge());
+            }
+            builder.put(totalName(topic, "size"), new AtomicLongGauge());
+            builder.put(totalName(topic, "offset"), new AtomicLongGauge());
+            builder.put(totalName(topic, "lag"), new AtomicLongGauge());
+        });
+
+        if (!badTopics.isEmpty()) {
+            throw new IllegalArgumentException(
+                    String.format("bad topic%s: '%s'", badTopics.size() != 1 ? "s" : "", badTopics));
+        }
+
+        metricMap = builder.build();
+
+        exec = Executors.newSingleThreadScheduledExecutor(
+                new ThreadFactoryBuilder()
+                        .setNameFormat(String.format("%s-offset-metrics-%%d", groupId))
+                        .build()
+        );
+    }
+
+    private String partitionName(final String topic, final int partition, final String name) {
+        return String.format("%s.%s.partition.%d.%s", metricPrefix, topic, partition, name);
+    }
+
+    private String totalName(final String topic, final String name) {
+        return String.format("%s.%s.total.%s", metricPrefix, topic, name);
+    }
+
+    @VisibleForTesting
+    Map<String, Metric> getMetrics() {
+        return metricMap;
+    }
+
+    public void start() {
+        metricRegistry.registerAll(() -> metricMap);
+        exec.scheduleAtFixedRate(this::poll, 0, pollPeriod.toMillis(), TimeUnit.MILLISECONDS);
+    }
+
+    public void stop() {
+        MetricSets.removeAll(metricRegistry, () -> metricMap);
+        try {
+            OTExecutors.shutdownAndAwaitTermination(exec, Duration.ofSeconds(5));
+        } catch (final InterruptedException e) {
+            LOG.error("worker thread shutdown interrupted", e);
+            Thread.currentThread().interrupt();
+        }
+        monitor.close();
+    }
+
+    @Override
+    public void close() {
+        stop();
+    }
+
+    private void poll() {
+        topics.forEach(topic -> {
+            try {
+                pollUnsafe(topic);
+            } catch (final Exception e) {
+                LOG.error("caught exception for topic {}", topic, e);
+            }
+        });
+    }
+
+    private void pollUnsafe(final String topic) {
+        final Map<Integer, Long> sizes = monitor.getTopicSizes(topic);
+        sizes.forEach((part, size) -> gauge(partitionName(topic, part, "size")).set(size));
+        gauge(totalName(topic, "size")).set(sumValues(sizes));
+
+        Map<Integer, Long> offsets = monitor.getGroupOffsets(groupId, topic);
```

Comment:
If you call getGroupOffsets first, wouldn't that eliminate the possibility of negative lag? It would bias towards larger lags. Also, call getGroupOffsets/getTopicSizes before doing any processing.

Reply:
That's true @enorrbinot.  As you say, however, the bias will still exist.  So it's not as compelling from that standpoint.  Further, it's a little better in the code to fetch the topic sizes first, so we can set those unconditionally.  Note how we do an early out a little bit later on.

What d' ya think?

Reply:
Agreed

Reply:
Ok, cool.